controllers/packetmachine_controller.go

```diff
@@ -236,13 +245,11 @@ func (r *PacketMachineReconciler) reconcile(ctx context.Context, machineScope *s
 			createDeviceReq.ControlPlaneEndpoint = controlPlaneEndpoint.Address
 		}
 
-		createDeviceReq.ExtraTags = tags
-
 		dev, err = r.PacketClient.NewDevice(createDeviceReq)
 
 		switch {
 		// TODO: find a better way than parsing the error messages for this.
-		case strings.Contains(err.Error(), " no available hardware reservations "):
+		case err != nil && strings.Contains(err.Error(), " no available hardware reservations "):
```

Comment:
Why is this non-fatal? (I could use comments)

Does the provider API return an error even though the device will be created, making this more of a provider warning than a fatal error? 
Is the HTTP response a 200 with an `errors` block, perhaps? If so that may be a better way to handle this than string matching.

Reply:
Updating the comment to provide more info, but if we do not special case this then Machine creation will fail after the first attempt and will not be retried (presence of ErrorMessage/ErrorReason in CAPI indicates a fatal error that should not be retried).

This was causing issues with users that wanted to run clusters on reserved hardware without having additional free reservations available during the upgrade, since attempts to create new Machines before the reservations were made available by Machines that were being deleted weren't immediately available.

Reply:
> Now in this block, we are going to try to adopt any existing device that carries the tags we would have placed in our previous attempt to create this device.

> Is this because the create call can error before returning the UUID (perhaps a timeout), somehow despite the error the device creation may succeed, and we want to try to salvage the creation?

It mainly means that we haven't recorded the providerID for any reason. That could be:
- k8s api call to patch the resource to set the providerID failed
- controller crashed, preventing reconciliation from completing
- api call to EM API to create device succeeded, but returned an error for some reason (could be timeout)

Thinking through this more, I should also update `reconcileDelete` to account for these use cases as well to avoid orphaning resources in the same cases as above.